src/realm/object_converter.cpp

```diff
@@ -253,29 +246,192 @@ void InterRealmValueConverter::copy_dictionary(const Obj& src_obj, Obj& dst_obj,
+            realm::Set<Mixed> src_set{src_obj, m_src_col};
+            realm::Set<Mixed> dst_set{dst_obj, m_dst_col};
+            // sets cannot be nested, so we just need to copy the values
+            copy_set(src_set, dst_set, update_out);
+        }
+        else if (src_mixed.is_type(type_Dictionary)) {
+            dst_obj.set_collection(m_dst_col, CollectionType::Dictionary);
+            Dictionary src_dict{src_obj, m_src_col};
+            Dictionary dst_dict{dst_obj, m_dst_col};
+            handle_dictionary_in_mixed(src_dict, dst_dict, update_out);
+            copy_dictionary(src_dict, dst_dict, update_out);
+        }
+        else {
+            InterRealmValueConverter::ConversionResult converted_src;
+            auto dst_mixed = dst_obj.get_any(m_dst_col);
+            if (cmp_src_to_dst(src_mixed, dst_mixed, &converted_src, update_out)) {
+                if (converted_src.requires_new_embedded_object) {
+                    Obj new_embedded = dst_obj.create_and_set_linked_object(m_dst_col);
+                    track_new_embedded(converted_src.src_embedded_to_check, new_embedded);
+                }
+                else {
+                    dst_obj.set_any(m_dst_col, converted_src.converted_value);
+                }
             }
-            else {
-                dst_obj.set_any(m_dst_col, converted_src.converted_value);
+        }
+    }
+}
+
+void InterRealmValueConverter::handle_list_in_mixed(const Lst<Mixed>& src_list, Lst<Mixed>& dst_list,
+                                                    bool* update_out)
+{
+    // utility functions
+    auto handle_set_copy = [&, this](size_t i) {
+        if (i >= dst_list.size())
+            dst_list.insert_collection(i, CollectionType::Set);
+        auto n_src_set = src_list.get_set(i);
+        auto n_dst_set = dst_list.get_set(i);
+        copy_set(*n_src_set, *n_dst_set, update_out);
+    };
+
+    auto handle_list_copy = [&, this](size_t i) {
+        if (i >= dst_list.size())
+            dst_list.insert_collection(i, CollectionType::List);
+        auto n_src_list = src_list.get_list(i);
+        auto n_dst_list = dst_list.get_list(i);
+        handle_list_in_mixed(*n_src_list, *n_dst_list, update_out);
+        copy_list(*n_src_list, *n_dst_list, update_out);
+    };
+
+    auto handle_dict_copy = [&, this](size_t i) {
+        if (i >= dst_list.size())
+            dst_list.insert_collection(i, CollectionType::Dictionary);
+        auto n_src_dict = src_list.get_dictionary(i);
+        auto n_dst_dict = dst_list.get_dictionary(i);
+        handle_dictionary_in_mixed(*n_src_dict, *n_dst_dict, update_out);
+        copy_dictionary(*n_src_dict, *n_dst_dict, update_out);
+    };
+
+    // main algorithm
+    auto n = src_list.size();
+    for (size_t i = 0; i < n; ++i) {
+        auto any = src_list.get_any(i);
+        if (any.is_type(type_List)) {
+            if (i < dst_list.size() && check_collection_in_mixed_mismatch(any, dst_list.get_any(i), type_List))
```

Comment:
There shouldn't be any conditionals like this in this object_converter class; the code here is simply meant to always change "dst" to be the same as "src". The recovery needs to happen while replaying the local changesets and that code is handled by the `RecoveryResolver`. The reason this is important is because during client reset, the changes that happen within this `InterRealmValueConverter` class are not part of the client history (ie this part of the history is discarded). The consequence of this, is that the client may appear to have recovered correctly but actually it did not create any sync history for the state it is now in; this will make a divergence between the server and the client which can lead to weird errors.

Reply:
I think that check (and also the others) are stale code, they don't longer apply. I will remove them.